PyInstaller/building/imphook.py

```diff
@@ -15,21 +15,73 @@
 import glob
 import os.path
 import re
-import sys
-import warnings
 
+from collections import OrderedDict
 from .. import log as logging
 from .utils import format_binaries_and_datas
 from ..compat import expand_path
-from ..compat import importlib_load_source, is_py2
-from ..utils.misc import get_code_object
+from ..compat import importlib_load_source
 from .imphookapi import PostGraphAPI
-from ..lib.modulegraph.modulegraph import GraphError
 
 logger = logging.getLogger(__name__)
 
 
-class HooksCache(dict):
+# Reproducible freeze: OrderedDict ensures hooks are always applied in the same
+# order.
+class ExcludedImports(OrderedDict):
+    """
+    Dictionary mapping for hook attribute 'excludedimports'.
+    'excludedimports' is a list of Python module names that PyInstaller
+    should not detect as dependency of some module names.
+
+    Excludedimports attribute is parsed from all post-import hooks
+    before analyzing any Python code. This ensures that all excluded
+    modules are checked in the right time when they get imported.
+    """
+    def load_hooks(self, hooks_dict):
+        """
+        Parse `excludedimports` from hook files.
+        """
+        logger.info('Loading excluded imports...')
+        self._parser = re.compile('excludedimports = (.+)$', flags=re.MULTILINE)
+        # TODO find out better way how to obtain 'excludedimports' from hooks.
+        for modname, filenames in hooks_dict.items():
+            # TODO parse all files not just first one.
+            code = self._parse_code(filenames[0])
+            if not code:  # hook does not contain 'excludedimports'
+                continue
+            # Evaluate code and update self dict.
+            excl_imports = eval(code)
```

Comment:
Bound to make trouble: This will crash as soon as someone write an expression revering to some variable.